### ISSUE TITLE
Error while using remora model

### ISSUE THREAD

**Issue by touala** [nanoporetech/remora#82]
Hello,

Thanks for your work on `remora`. Unfortunately, I have the following issue while trying to apply a model I trained:
```sh
remora -v
# Remora version: 2.1.1
remora infer from_pod5_and_bam barcode11.pod5 small.sorted.bam \
  --model train_results_2.1.1/model_best.pt \
  --reference-anchored \
  --out-bam out.remora.bam \
  --log-filename out.remora.log
# Indexing BAM by read id: 75 Reads [00:00, 6621.89 Reads/s]
# [09:59:05] Extracting read IDs from POD5
# [09:59:06] Found 75 BAM records, 261923 POD5 reads, and 75 in common
# Inferring mods:   0%|                                                                                                                      | 0/75 [00:00<?, ? Reads/s]
# double free or corruption (out)
# Typed *Ctrl-C*
# Process PrepBatches_filler:
# Inferring mods:   0%|         
```

```sh
$ cat out.remora.log
DEBUG [09:59:04:MainProcess:MainThread:log.py:67] Command: """/path/to/remora_latest/venv/bin/remora infer from_pod5_and_bam barcode11.pod5 small.sorted.bam --model train_results_2.1.1/model_best.pt --reference-anchored --out-bam out.remora.bam --log-filename out.remora.log"""
DEBUG [09:59:04:MainProcess:MainThread:model_util.py:423] Using torchscript model
DEBUG [09:59:04:MainProcess:MainThread:model_util.py:401] Loaded Remora model attrs
         creation_date : 05/29/2023, 22:02:07
    kmer_context_bases : [4, 4]
         chunk_context : [50, 50]
             base_pred : False
             mod_bases : x
        reverse_signal : False
    base_start_justify : False
                offset : 0
          model_params : {'size': 64, 'kmer_len': 9, 'num_out': 2}
            num_motifs : 1
            doc_string : Nanopore Remora model
         model_version : 3
        mod_long_names : ['xxx']
              kmer_len : 9
             chunk_len : 100
                motifs : [('A', 0)]
              can_base : A
                 motif : ('A', 0)
          alphabet_str : loaded modified base model to call (alt to A): x=xxx
       sig_map_refiner : Loaded 0-mer table with 0 central position. Rough re-scaling will be executed.

[09:59:05] Extracting read IDs from POD5
[09:59:06] Found 75 BAM records, 261923 POD5 reads, and 75 in common
DEBUG [09:59:06:ExtractSignal_filler:MainThread:util.py:436] Starting ExtractSignal background filler
DEBUG [09:59:06:ExtractSignal_filler:MainThread:io.py:348] Reading from POD5 at barcode11.pod5
DEBUG [09:59:06:AddAlignments_0:MainThread:util.py:407] Starting AddAlignments worker
DEBUG [09:59:06:PrepBatches_0:MainThread:util.py:407] Starting PrepBatches worker
DEBUG [09:59:08:ExtractSignal_filler:MainThread:io.py:360] Completed pod5 signal worker
DEBUG [09:59:08:ExtractSignal_filler:MainThread:io.py:389] Completed signal worker
DEBUG [09:59:08:ExtractSignal_filler:MainThread:util.py:447] Completed ExtractSignal background filler
DEBUG [09:59:08:AddAlignments_0:MainThread:util.py:431] Completed AddAlignments worker
DEBUG [10:00:23:MainProcess:MainThread:util.py:514] MultitaskMap PrepBatches interrupted
```

I encounter the same error while trying to run on a GPU with `--device 0`. 

[Here](https://github.com/nanoporetech/remora/files/11610441/remora_issue.packages.txt) is `pip list` output. I see some parallel with #67  but it seems to be a different error. It was installed with:

```sh
git clone https://github.com/nanoporetech/remora.git remora_latest
python3.8 -m remora_latest/venv --prompt remora --copies venv
source remora_latest/venv/bin/activate
pip install -e remora_latest/[tests]
```

I've tried to use this remora model with dorado `v0.3.0` but got `Unknown modified base abbreviation` which is not fixed with `--modified-bases`. With bonito `v0.7.0` the model is loading but I get the following error after a couple of seconds:
```sh
bonito basecaller -v --modified-base-model train_results_2.1.1/model_best.pt --modified-device cuda:0 dna_r10.4.1_e8.2_400bps_sup@v4.1.0 --device cuda:0 pod5_dir  > pod5_dir/basecalls.bonito.bam
> reading pod5
> outputting unaligned bam
> loading model dna_r10.4.1_e8.2_400bps_sup@v4.1.0
> model basecaller params: {'batchsize': 96, 'chunksize': 10000, 'overlap': 500, 'quantize': None}
> loading modified base model
> loaded modified base model to call (alt to A): x=xxx
> calling:   0%|                                                      | 0/40772 [00:00<?, ? reads/s]free(): invalid next size (normal)
Abandon (core dumped)
```

Am I missing something?

Thanks.

**Comment by touala** [comment]
Thanks for the quick turnaround. In `remora dataset prepare`, I've used `--refine-rough-rescale` but didn't provide `--refine-kmer-level-table <file>`. Sounds like this could be the issue. I'm rerunning the training and I will get back to you.

**Comment by touala** [comment]
That was it. Now I see the issue from #67 but I'll roll back to `torch==1.13.1` and retrain a model for now. I'm assuming GPU inference is faster than CPU one, is it the case?

Also, I don't know how I missed it but the issue was already reported with a fix in #77, sorry about that. Although after reading this post, I'm wondering how `--refine-rough-rescale` affect trained models performances. Does it enable a trade-off toward faster preprocessing but with lower model prediction accuracy? It seems that `--refine-scale-iters <int>` also enable more computationally intensive steps, could you comment on potential model accuracy gains?

Best.

**Comment by marcus1487** [comment]
The issue with errors on infer should be resolved with the latest release 2.1.2. 

For the refinement options, the rough rescale option is highly recommended. The compute cost for this step is very minimal and results in a significant model improvement. The refine iters option is much more experimental and computationally expensive with minimal extra benefit in most tested cases. The refine iters algorithms are not implemented in dorado as they have not proven sufficiently beneficial to this point. The refine iters option performs potentially more precise rescaling and signal assignment options, but this does not seem to robustly improve results in practice. Hopefully this helps some in understanding the best practices in remora.

**Comment by touala** [comment]
Thank you Marcus for your comments. I'll try `2.1.2`. Best.
